Mobile reconnection doc (#1184)
spatialos/gdk-for-unity: docs/modules/mobile/reconnection-logic.md

<%(TOC)%>

# Reconnection logic

> Mobile applications present new challenges in terms of handling client connections. A client may disconnect due to an unstable connection or when the application is put into a background state by the OS. We want to dive a bit deeper into the causes of a disconnect and how you can handle it in your game.

Reviewer: ```suggestion
Mobile applications present new challenges in terms of handling client connections. A client may disconnect due to an unstable connection or when the application is put into a background state by the OS. In this page, we will dive a bit deeper into the causes of a disconnect and discuss how you can handle it in your game.
```

## Why clients disconnect

### Unstable connection 

> When creating a mobile application, you need to ensure that [client-workers]({{urlRoot}}/reference/glossary#client-worker) are able to run the game even with an unstable connection. To achieve this, you can implement connection handling to ensure that those client-workers are able to reconnect.

Reviewer: I don't really know what "connection handling" is

If a client-worker's connection is too unstable, they might not be able to connect to your game or might disconnect at any time. You should provide a connection flow that allows them to gracefully disconnect and handle reconnection when they have a more stable connection.

### Pausing of applications

> When pausing an application or putting it into the background, there are two possible scenarios that can cause a disconnect:

Reviewer: ```suggestion
When pausing an application or putting it into the background, there are at least two possible scenarios that can cause a disconnect:
```

Speculatively?

#### 1. The device OS closes the application.

> If the application has been closed, the user has to restart the application. The user will be brought back to the start screen the next time the application is opened and from there on, the application can connect and load any data necessary.

Reviewer: In this scenario, you lean more into the player identity side of things right? Do we want to touch upon that?

Author: I would say that this would need it's own documentation

#### 2. The client stops sending data. 

> This scenario is a bit more tricky. The application is still alive, however the OS may decide to not run the game or send any data while the application is in the background. [SpatialOS]({{urlRoot}}/reference/glossary#spatialos-runtime) will close the connection if it doesn't receive any messages from a client-worker for a period of time. You need to add additional reconnection logic to your game to handle this scenario. 

Reviewer: ```suggestion
This scenario is a bit more tricky. The application is still alive, however the OS may decide to not run the game or send any data while the application is in the background. [SpatialOS]({{urlRoot}}/reference/glossary#spatialos-runtime) will close the connection if it doesn't receive any messages from a client-worker for a period of time. 

When the application is brought back to the foreground, it will act as if the connection was dropped. You need to add additional reconnection logic to your game to handle this scenario. 
```

This is handled basically the same as the Unstable Connection section right? Do we want to reference that?

In the end, this is a very game-specific question that depends entirely on what you want to offer to your users. For example, the FPS Starter Project implements the simplest solution: a disconnect takes you back to the start screen, from where you can reconnect to the game.

> ## How to detect a disconnect

Reviewer 2: I think this section is more of a "what types of disconnect", which would then go into how they are detected

[Heartbeating]({{urlRoot}}/modules/player-lifecycle/heartbeating) is a technique used to verify that your client is still connected. If there are too many failed heartbeats, there are two ways a client-worker may be seen as disconnected:

### 1. SpatialOS believes you disconnected

> To ensure that a worker is still connected, the worker has to send messages to SpatialOS at a set interval. This indicates that the worker is still alive. If SpatialOS doesn't receive any message for a while, it will close the connection to that worker. 

Reviewer: ```suggestion
To ensure that a worker is still connected, the worker has to send messages to SpatialOS at a set interval. This indicates that the worker is still alive. If SpatialOS doesn't receive any messages for a while, it will close the connection to that worker. 
```

<%(TOC)%>

# Reconnection logic

> Mobile applications present new challenges in terms of handling client connections. A client may disconnect due to unstable connections or when the application is put into a background state by the OS. We want to dive a bit deeper into the causes of a disconnect and how you can handle it in your game.

Reviewer: ```suggestion
Mobile applications present new challenges in terms of handling client connections. A client may disconnect due to an unstable connection or when the application is put into a background state by the OS. We want to dive a bit deeper into the causes of a disconnect and how you can handle it in your game.
```

## Why clients disconnect

### Unstable connection 

> When creating a mobile application, you need to ensure that [client-workers]({{urlRoot}}/reference/glossary#client-worker) are able to run the game even with an unstable connection.

Reviewer: How do I do this?

Author: added something, ptal

> If a client-worker's connection is too unstable, they might not be able to connect to your game or might disconnect at any time. You should provide a connection flow that allows them to disconnect properly, and handle reconnection as soon as they obtain better reception.

Reviewer: ```suggestion
If a client-worker's connection is too unstable, they might not be able to connect to your game or might disconnect at any time. You should provide a connection flow that allows them to gracefully disconnect and handle reconnection when they have more stable connection.
```

### Pausing of applications

When pausing an application or putting it into the background, there are two possible scenarios that can cause a disconnect:

#### 1. The device OS closes the application.

If the application has been closed, the user has to restart the application. The user will be brought back to the start screen the next time the application is opened and from there on, the application can connect and load any data necessary.

#### 2. The client stops sending data. 

> This scenario is a bit more tricky. The application is still alive, however the OS may decide to not run the game or send any data while the application is in the background. [SpatialOS]({{urlRoot}}/reference/glossary#spatialos-runtime) will close the connection, if it doesn't receive any messages from a client-worker in a while. You need to add additional reconnection logic to your game to be able to handle this scenario. 

Reviewer: ```suggestion
This scenario is a bit more tricky. The application is still alive, however the OS may decide to not run the game or send any data while the application is in the background. [SpatialOS]({{urlRoot}}/reference/glossary#spatialos-runtime) will close the connection if it doesn't receive any messages from a client-worker for a period of time. You need to add additional reconnection logic to your game to handle this scenario. 
```

> In the end this is a very game-specific question that depends entirely on what you want to offer to your users. The FPS Starter Project implements the simplest solution: a disconnect takes you back to the start screen, from where you can reconnect to the game.

Reviewer: ```suggestion
In the end, this is a very game-specific question that depends entirely on what you want to offer to your users. For example, the FPS Starter Project implements the simplest solution: a disconnect takes you back to the start screen, from where you can reconnect to the game.
```

## How to detect a disconnect

[Heartbeating]({{urlRoot}}/modules/player-lifecycle/heartbeating) is a technique used to verify that your client is still connected. If there are too many failed heartbeats, there are two ways a client-worker may be seen as disconnected:

### 1. SpatialOS believes you disconnected

> To ensure that a worker is still connected, the worker has to send messages to SpatialOS every now and thne. This indicates that the worker is still alive. If SpatialOS doesn't receive any message for a while, it will close the connection to that worker. 

Reviewer: ```suggestion
To ensure that a worker is still connected, the worker has to send messages to SpatialOS at a set interval. This indicates that the worker is still alive. If SpatialOS doesn't receive any message for a while, it will close the connection to that worker. 
```

> Whenever the connection gets closed on a worker, that worker receives a `DisconnectOp` object. It contains the reason behind the disconnection and triggers a disconnect event in the GDK.

Reviewer: ```suggestion

Whenever the connection gets closed on a worker, that worker receives a `DisconnectOp` object. It contains the reason behind the disconnection and triggers a disconnect event in the GDK.
```

> You can listen for this event using the `Worker` object inside the `WorkerConnector` class in order to perform any kind of disconnection logic necessary.

Reviewer: ```suggestion
You can listen for this event using the `Worker` object inside the `WorkerConnector` class in order to perform your disconnection logic.
```

```csharp
namespace YourGame
{
    public class MobileClientWorkerConnector : WorkerConnector
    {
    	...

        protected override void HandleWorkerConnectionEstablished()
        {
             Worker.OnDisconnect += OnDisconnected;
        }

        private void OnDisconnected(string reason)
        {
            Worker.LogDispatcher.HandleLog(LogType.Log, new LogEvent($"Worker disconnected")
                .WithField("WorkerId", Worker.WorkerId)
                .WithField("Reason", reason));
        }
    }
}
```

### 2. The player lifecycle module believes you disconnected 

> Our [Player Lifecycle module]({{urlRoot}}/modules/player-lifecycle/overview) allows you to easily create players and also implements heartbeating to delete the player entity of any unresponsive client-workers. In mobile applications, this can result in entities being destroyed even though the worker is still connected to SpatialOS. If this happens, that client-worker won't receive a `DisconnectOp`.

Reviewer: ```suggestion
Our [Player Lifecycle module]({{urlRoot}}/modules/player-lifecycle/overview) allows you to easily create players and uses the heartbeating technique to detect and delete the player entity of any unresponsive client-workers. In mobile applications, this can result in entities being destroyed even though the worker is still connected to SpatialOS. If this happens, that client-worker won't receive a `DisconnectOp`.
```

Reviewer: > In mobile applications, this can result in entities being destroyed even though the worker is still connected to SpatialOS.

Under what conditions can this happen? I.e. - client level heartbeats fail, but the underlying SpatialOS ones don't

Author: yeah, will add a sentence about that

> When the player entity gets deleted, the client-worker might end up with no entities to be authoritative over and therefore unable to check out any entities. The client-worker's world will appear to be empty.

Reviewer: I think more generally, this leaves the client in a bad (but usually recoverable) state.

Author: reworded it a bit

<%(TOC)%>

# Reconnection logic

> Mobile applications present new challenges in terms of handling client connections. A client may disconnect due to unstable connections or when pausing the application. We want to dive a bit deeper into the causes of a disconnect and how you can handle it in your game.

Reviewer: ```suggestion
Mobile applications present new challenges in terms of handling client connections. A client may disconnect due to unstable connections or when the application is put into a background state by the OS. We want to dive a bit deeper into the causes of a disconnect and how you can handle it in your game.
```

Reviewer: To capture the specific behaviour of mobile devices vs just application is paused (which feels like a user-level action rather than a system one)

## Why clients disconnect

> ### Unstable connection / cellular reception

Reviewer: We say cellular reception here but then never reference it in the section below

Author: removing it

When creating a mobile application, you need to ensure that clients are able to play the game even with an unstable connection.

If a client's connection is too unstable, they might not be able to connect to your game or might disconnect at any time. You should provide a connection flow that allows them to disconnect properly, and handle reconnection as soon as they obtain better reception.

### Pausing of applications

When pausing an application or putting it into the background, there are two possible scenarios that can cause a disconnect:

#### 1. The device OS closes the application.

If the application has been closed, the user has to restart the application. The user will be brought back to the start screen the next time the application is opened and from there on, the application can connect and load any data necessary.

> #### 2. The game stops sending data. If this happens for too long, SpatialOS will close the connection.

Reviewer: Why does the game stop sending data?

Author: added some explanation in the section

This scenario is a bit more tricky. The application is still alive, however the next time the user opens the application, SpatialOS will have already closed the connection. You need to add additional reconnection logic to your game to be able to handle this scenario. 

> In the end this is a very game-specific question that depends entirely on what you want to offer to your users. The FPS Starter Project implements the simplest solution: a disconnect takes you back to the start screen, from where you can reconnect to continue the game.

Reviewer: Continue is a bit incorrect here since you aren't the same player right?

Author: depends on how the game is designed, but will change it

## How to detect a disconnect

[Heartbeating]({{urlRoot}}/modules/player-lifecycle/heartbeating) is a technique used to verify that your client is still connected. If there are too many failed heartbeats, there are two ways your client may be seen as disconnected:

### 1. SpatialOS believes you have disconnected

> If SpatialOS disconnects you, you receive a `DisconnectOp`. This object contains the reason behind the disconnection and triggers a disconnect event in the GDK.

Reviewer: You actually get this op regardless of how you disconnect. Its generated by the Worker SDK when the connection closes.

You can listen for this event using the `Worker` object inside the `WorkerConnector` class in order to perform any kind of disconnection logic necessary.

```csharp
namespace YourGame
{
    public class MobileClientWorkerConnector : WorkerConnector
    {
    	...

        protected override void HandleWorkerConnectionEstablished()
        {
             Worker.OnDisconnect += OnDisconnected;
        }

        private void OnDisconnected(string reason)
        {
            Worker.LogDispatcher.HandleLog(LogType.Log, new LogEvent($"Worker disconnected")
                .WithField("WorkerId", Worker.WorkerId)
                .WithField("Reason", reason));
        }
    }
}
```

> ### 2. The server-worker believes you have disconnected

Reviewer: This is something that only happens with the Player Lifecycle feature module. Server workers don't believe anything about the state of the client connection unless you design it to do so.

Author: true, might need to rephrase this a bit as it's definitely important to note that this can happen, if you use the player lifecycle module
